PEP 649: Improve description of semantics and mark as accepted (#3138)
python/peps: pep-0649.rst

PEP: 649
Title: Deferred Evaluation Of Annotations Using Descriptors
Author: Larry Hastings <larry@hastings.org>
Status: Accepted
Type: Standards Track
Topic: Typing
Content-Type: text/x-rst
Created: 11-Jan-2021
Python-Version: 3.13

> Post-History: 11-Jan-2021, 11-Apr-2021, 19-Apr-2023

Reviewer: ```suggestion
Post-History: 11-Jan-2021, 11-Apr-2021, 19-Apr-2023
Resolution: https://discuss.python.org/t/a-massive-pep-649-update-with-some-major-course-corrections/25672/62
```

(https://peps.python.org/pep-0001/#pep-header-preamble)

Author: Good call!  FWIW I'm actually gonna go with [the other post from the SC announcing they were accepting 649.](https://discuss.python.org/t/pep-649-deferred-evaluation-of-annotations-tentatively-accepted/21331/43)  This was their own thread about acceptance of 649 so it seemed more on point somehow.

PEP: 649
Title: Deferred Evaluation Of Annotations Using Descriptors
Author: Larry Hastings <larry@hastings.org>

> Status: Accepted

Reviewer 2: Pending SC review — please mark this comment as resolved IFF and when the SC accepts the changes (so it blocks accidental merge otherwise). Also marked as `DO-NOT-MERGE` (which is currently not actually checked by the CIs on this repo; Draft would be a more robust solution but could send the wrong signal about others reviewing it).

Type: Standards Track
Topic: Typing
Content-Type: text/x-rst
Created: 11-Jan-2021
Python-Version: 3.13

> Post-History: 11-Jan-2021, 11-Apr-2021, 19-Apr-2023, 10-May-2023

Reviewer 2: ```suggestion
Post-History: `11-Jan-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/5QMMCRF4HTRRNJV56CGHVI5GRHVBDGQO/>`__,
              `12-Apr-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/QSASX6PZ3LIIFIANHQQFS752BJYFUFPY/>`__,
              `18-Apr-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/WUZGTGE43T7XV3EUGT6AN2N52OD3U7AE/>`__,
              `09-Aug-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/2MEOWHCVDLPABOBLYUGRXVOOOBYOLLU6/>`__,
              `20-Oct-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/SZLWVYV2HPLU6AH7DOUD7DWFUGBJGQAY/>`__,
              `20-Oct-2021 <https://discuss.python.org/t/type-annotations-pep-649-and-pep-563/11363>`__,
              `17-Nov-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/VIZEBX5EYMSYIJNDBF6DMUMZOCWHARSO/>`__,
              `15-Mar-2022 <https://discuss.python.org/t/finding-edge-cases-for-peps-484-563-and-649-type-annotations/14314>`__,
              `23-Nov-2022 <https://discuss.python.org/t/pep-649-deferred-evaluation-of-annotations-tentatively-accepted/21331>`__,
              `07-Feb-2023 <https://discuss.python.org/t/two-polls-on-how-to-revise-pep-649/23628>`__,
              `11-April-2023 <https://discuss.python.org/t/a-massive-pep-649-update-with-some-major-course-corrections/25672>`__,
```

After a bunch of post archeology on Discourse and Python-Dev, I've managed to, I believe, fairly accurately reconstruct the (looooooong, on this one) history of discussion threads for this PEP.

I've only listed threads that had as its core topic the discussion of PEP 649 (including comparison with the alternative PEP 563, but with 649 as a main focus), and limited it to those that were posted by either @larryhastings or the SC to gather PEP feedback (with the exception of [one thread](https://mail.python.org/archives/list/python-dev@python.org/thread/2MEOWHCVDLPABOBLYUGRXVOOOBYOLLU6/) by Eric V. Smith that was a continuation of Larry's previous PEP discussion thread, posted in apparent coordination with the same and was indeed the de-facto next iteration of the PEP discussion thread).

This included the threads that appeared to corresponding to the existing bare dates already listed (it seems there isn't a new one for this PEP update, so that can be added later if a new one is posted rather than posting to one of the already-listed threads).

@larryhastings , if you think something shouldn't be included or have something else that you think should be, just let me know of course!

Author: Goodness me!  You're doing yeoman's work here.  Thanks, CAM!

I'm catching a plane tonight and will have a busy day or two.  I'll try to review your list here before the weekend.  On the other hand, this change is low-stakes; if we missed one, or had an extra one we shouldn't, we can just update the list.  I don't think changing that even after the PEP is accepted is a problem.  So I'd be satisfied merging your list regardless.

> Post-History: 11-Jan-2021, 11-Apr-2021, 19-Apr-2023, 10-May-2023

Reviewer 2: ```suggestion
Post-History: `11-Jan-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/5QMMCRF4HTRRNJV56CGHVI5GRHVBDGQO/>`__,
              `12-Apr-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/QSASX6PZ3LIIFIANHQQFS752BJYFUFPY/>`__,
              `18-Apr-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/WUZGTGE43T7XV3EUGT6AN2N52OD3U7AE/>`__,
              `09-Aug-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/2MEOWHCVDLPABOBLYUGRXVOOOBYOLLU6/>`__,
              `20-Oct-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/SZLWVYV2HPLU6AH7DOUD7DWFUGBJGQAY/>`__,
              `20-Oct-2021 <https://discuss.python.org/t/type-annotations-pep-649-and-pep-563/11363>`__,
              `17-Nov-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/VIZEBX5EYMSYIJNDBF6DMUMZOCWHARSO/>`__,
              `15-Mar-2022 <https://discuss.python.org/t/finding-edge-cases-for-peps-484-563-and-649-type-annotations/14314>`__,
              `23-Nov-2022 <https://discuss.python.org/t/pep-649-deferred-evaluation-of-annotations-tentatively-accepted/21331>`__,
              `07-Feb-2023 <https://discuss.python.org/t/two-polls-on-how-to-revise-pep-649/23628>`__,
              `11-April-2023 <https://discuss.python.org/t/a-massive-pep-649-update-with-some-major-course-corrections/25672>`__,
```

After a bunch of post archeology on Discourse and Python-Dev, I've managed to, I believe, fairly accurately reconstruct the (looooooong, on this one) history of discussion threads for this PEP.

I've only listed threads that had as its core topic the discussion of PEP 649 (including comparison with the alternative PEP 563, but with 649 as a main focus), and limited it to those that were posted by either @larryhastings or the SC to gather PEP feedback (with the exception of [one thread](https://mail.python.org/archives/list/python-dev@python.org/thread/2MEOWHCVDLPABOBLYUGRXVOOOBYOLLU6/) by Eric V. Smith that was a continuation of Larry's previous PEP discussion thread, posted in apparent coordination with the same and was indeed the de-facto next iteration of the PEP discussion thread).

This included the threads that appeared to corresponding to the existing bare dates already listed (it seems there isn't a new one for this PEP update, so that can be added later if a new one is posted rather than posting to one of the already-listed threads).

@larryhastings , if you think something shouldn't be included or have something else that you think should be, just let me know of course!

> Post-History: 11-Jan-2021, 11-Apr-2021, 19-Apr-2023, 10-May-2023

Reviewer 2: ```suggestion
Post-History: `11-Jan-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/5QMMCRF4HTRRNJV56CGHVI5GRHVBDGQO/>`__,
              `12-Apr-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/QSASX6PZ3LIIFIANHQQFS752BJYFUFPY/>`__,
              `18-Apr-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/WUZGTGE43T7XV3EUGT6AN2N52OD3U7AE/>`__,
              `09-Aug-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/2MEOWHCVDLPABOBLYUGRXVOOOBYOLLU6/>`__,
              `20-Oct-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/SZLWVYV2HPLU6AH7DOUD7DWFUGBJGQAY/>`__,
              `20-Oct-2021 <https://discuss.python.org/t/type-annotations-pep-649-and-pep-563/11363>`__,
              `17-Nov-2021 <https://mail.python.org/archives/list/python-dev@python.org/thread/VIZEBX5EYMSYIJNDBF6DMUMZOCWHARSO/>`__,
              `15-Mar-2022 <https://discuss.python.org/t/finding-edge-cases-for-peps-484-563-and-649-type-annotations/14314>`__,
              `23-Nov-2022 <https://discuss.python.org/t/pep-649-deferred-evaluation-of-annotations-tentatively-accepted/21331>`__,
              `07-Feb-2023 <https://discuss.python.org/t/two-polls-on-how-to-revise-pep-649/23628>`__,
              `11-April-2023 <https://discuss.python.org/t/a-massive-pep-649-update-with-some-major-course-corrections/25672>`__,
```

After a bunch of post archeology on Discourse and Python-Dev, I've managed to, I believe, fairly accurately reconstruct the (looooooong, on this one) history of discussion threads for this PEP.

I've only listed threads that had as its core topic the discussion of PEP 649 (including comparison with the alternative PEP 563, but with 649 as a main focus), and limited it to those that were posted by either @larryhastings or the SC to gather PEP feedback (with the exception of [one thread](https://mail.python.org/archives/list/python-dev@python.org/thread/2MEOWHCVDLPABOBLYUGRXVOOOBYOLLU6/) by Eric V. Smith that was a continuation of Larry's previous PEP discussion thread, posted in apparent coordination with the same and was indeed the de-facto next iteration of the PEP discussion thread).

This included the threads that appeared to corresponding to the existing bare dates already listed (it seems there isn't a new one for this PEP update, so that can be added later if a new one is posted rather than posting to one of the already-listed threads).

@larryhastings , if you think something shouldn't be included or have something else that you think should be, just let me know of course!

Implementation
**************

Observed semantics for annotations expressions
==============================================

For any object ``o`` that supports annotations,
``o.__annotations__`` will produce an identical
annotations dict under this PEP as it would
under "stock" semantics, provided that none of
the names referenced in the annotations expressions
are subsequently rebound.

In particular, when this PEP is active, name
resolution in annotations must be *identical*

> to stock semantics.

Author: I'm happy to reword or even strike the paragraph.  My goal with this paragraph was to make an explicit guarantee for the reader that name resolution would be identical.  Yes, it's redundant with the previous paragraph, but sometimes repeating yourself adds emphasis, which here I thought would be comforting to readers familiar with the Python annotations saga.

Maybe we could fold it in to the previous paragraph?  Here's a bit of a re-wording:

> For any object `o` that supports annotations,
> provided that all names evaluated in the annotations expressions
> are bound before `o` is defined and never subsequently rebound,
> `o.__annotations__`  will produce an identical annotations dict both
> when "stock" semantics are active and when this PEP is active.
> In particular, name resolution will be performed identically in
> both scenarios.

> to stock semantics.

Author: I fleshed out that section a little bit more, to mention the caching, and to mention that the calculation is delayed until o.__annotations__ is first requested, which means

* names evaluate to their *current* value as of that moment, and
* if evaluating the annotation expressions raises an exception, the exception is raised as of that moment.

This'll be in the next update to the PR, please re-review.  Thanks!

Implementation
**************

Observed semantics for annotations expressions
==============================================

For any object ``o`` that supports annotations,
provided that all names evaluated in the annotations expressions
are bound before ``o`` is defined and never subsequently rebound,
``o.__annotations__`` will produce an identical annotations dict both
when "stock" semantics are active and when this PEP is active.
In particular, name resolution will be performed identically in
both scenarios.

When this PEP is active, the value of ``o.__annotations__``
won't be calculated until the first time ``o.__annotations__``
itself is evaluated.  All evaluation of the annotation expressions
is delayed until this moment, which also means that

* names referenced in the annotations expressions will use their
  *current* value at this moment, and
* if evaluating the annotations expressions raises an exception,
  that exception will be raised at this moment.

> Once ``o.__annotations__`` is calculated for the first time,

Author: Good call.  Done.

The callable stored in ``__annotate__`` must accept a
single required positional argument called ``format``,

> which will always be an ``int`` or an object interchangable

Reviewer 3: I rely on Google for checking my spelling, and Google says this has an e in it:
```suggestion
which will always be an ``int`` or an object interchangeable
```

Author: I thought the two spellings were interch... anyway I think we're removing this sentence fragment.

> with an ``int`` (e.g. ``enum.IntEnum``).  It must either

Reviewer 3: I'm not sure that "interchangeable" has a precise definition in terms of Python objects. If I were to be forced to assign a precise definition to it I would probably assume it means something like Liskov substitutability? But I think we could just say " or a subclass of int (e.g. enum.IntEnum)" and that would be more clear? `IntEnum` is a subclass of int.

I suppose technically you could pass in some other thing with an `__eq__` method that is willing to compare to ints, but a) if that's really what we mean here, we should say "comparable to" not "interchangeable with", and b) why should we suggest that such a thing is supported?

Author: I agree.  I went with

> will always be an `int` (or a subclass of `int`).

Seems sufficient, doesn't get lost in the weeds.  Thumbs up.

The callable stored in ``__annotate__`` must accept a
single required positional argument called ``format``,

> which will always be an ``int`` (or a subclas of ``int``).

Reviewer 4: Small typo:
```suggestion
which will always be an ``int`` (or a subclass of ``int``).
```

    annotations values should be provided.  Must be one of the
    following:

    ``inspect.VALUE`` (equivalent to the ``int`` constant ``1``)

        Values are the result of evaluating the annotation expressions.

>     ``inspect.SOURCE`` (equivalent to the ``int`` constant``2``)

Reviewer: Lint fix:

```suggestion
    ``inspect.SOURCE`` (equivalent to the ``int`` constant ``2``)
```

    annotations values should be provided.  Must be one of the
    following:

>     ``inspect.VALUE`` (equivalent to the ``int`` constant ``1``)

Reviewer 3: If we are putting these in an enum, should they really be aliased as top-level attributes of `inspect`? It always seemed slightly odd to me in terms of namespacing, since there are a ton of things in `inspect` that have nothing to do with annotations. Is it that painful for someone wanting to use these values to do `from inspect import AnnotationFormat` and then `AnnotationFormat.VALUE`?

Author: I think so.  `AnnotationFormat` is a lot to type.  Let me put it this way: this is a common enough idiom that enum explicitly supports it, via `@enum.global_enum`.

As a guy who uses `inspect.signature` more than the average bear, I often alias `inspect.Parameter.POSITIONAL_OR_KEYWORD` precisely because of how much *#%@! typing that is.

Reviewer 5: I believe `@global_enum` mostly exists for backward compatibility with previous module-level names. I won't insist on it but I'd also vote to get rid of the global names and use only the enum.

This PEP adds a new keyword-only parameter to these two functions,
``format``.  ``format`` specifies what format the values in the
annotations dict should be returned in.

> ``format`` accepts the following values, as defined the
> ``inspect`` module::

Reviewer 6: ```suggestion
``format`` accepts any member of the ``AnnotationFormat`` enumeration,
as defined in the ``inspect`` module::
```

Author: Well, if we want to really get down and exacting here, the `format` parameter of `inspect.get_annotations` accepts any value that *compares equal to* any member of the `AnnotationFormat` enumeration as defined in the `inspect` module.  For example, you can pass in `2`, and `get_annotations` will happily return a "forwardref" format dictionary.  So, how might you put that in a readable way?

Reviewer 3: We don't have to explicitly document support for everything that is technically possible. Are there strong use cases for someone using these public inspection methods to _not_ use the enum values?

Author: I don't.  I have a strong use case for the compiler not using the enum values, but the compiler isn't a person, and anyway that's an internal implementation detail.

Author: Oh, huh.  Now I get UI where I can comment directly.

I've revised this text.  Instead of redefining the enum for some reason, it just says

> The `format` parameter on these two functions accept the same values
> as the `format` parameter on the `__annotate__` magic method
> defined above; however, these `format` parameters also have a default
> value of `inspect.VALUE`.

    class AnnotationFormat(enum.IntEnum):
        VALUE = 1
        FORWARDREF = 2
        SOURCE = 3
    VALUE = AnnotationFormat.VALUE
    FORWARDREF = AnnotationFormat.FORWARDREF

>     SOURCE = AnnotationFormat.SOURCE

Reviewer 6: You could consider using ``enum.global_enum`` here perhaps ([reference](https://docs.python.org/3/library/enum.html#enum.global_enum)), though this would change the repr from ``<AnnotationFormat.VALUE: 1>`` to ``inspect.VALUE``.

Author: Thanks!  I hadn't heard of that.  I will probably do it.  But I don't think it has to go in the PEP.  I think I'd prefer to omit it from the PEP, actually, just to keep it light, breezy reading.

Author: I can't get it to work.  If I add a line above "class AnnotationFormat", indented at the same level as "class", that reads "@enum.global_enum", Sphinx:

`sphinx.errors.SphinxParallelError: docutils.utils.SystemMessage: /home/larry/src/python/peps/larrytrunk/pep-0649.rst:751: (ERROR/3) Unexpected indentation.`

If I screwed up the formatting, fair enough, but I don't think I did.  I think this is a Sphinx bug.

Can you get it to work?  I'm leaving it out for now.

Reviewer 6: > Can you get it to work?

I've gotten it to work, but can't do a suggestion via GH's interface -- do you mind if I push a commit to fix?

Author: Please do!

Reviewer 6: Done!

This PEP adds a new keyword-only parameter to these two functions,
``format``.  ``format`` specifies what format the values in the
annotations dict should be returned in.

> The ``format`` parameter on these two functions accept the same values

Reviewer 5: ```suggestion
The ``format`` parameters on these two functions accept the same values
```
(Or alternatively, add an s to "accept".)

Author: The 's' on accept is what I want, thanks.

Author: I concede it's a little weird that the sentence starts out talking about "The format parameter on these two functions", then after the semicolon talks about "these format parameters".  If you have a suggestion that makes it read better, by all means!

Reviewer 5: I think the current wording is fine.
